### PR TITLE
Use custom dispatcher for Array.prototype routines

### PR DESCRIPTION
**Pull request by rerobika**
Binary size reduction:
 - Intel: ~2.5KB (gcc-7.3)
 - Arm: ~1.2KB (arm-linux-gnueabi-gcc-7.3)

JerryScript-DCO-1.0-Signed-off-by: Robert Fancsik frobert@inf.u-szeged.hu

### CODE REVIEW

**Comment by rerobika** [comment]
Performance results:

Benchmark | Perf (sec) |
----: | ----: | 
3d-cube.js | 0.790 -> 0.788 : +0.224% | 
3d-raytrace.js | 0.950 -> 0.954 : -0.359% | 
access-binary-trees.js | 0.528 -> 0.528 : +0.106% | 
access-fannkuch.js | 1.913 -> 1.914 : -0.056% | 
access-nbody.js | 1.000 -> 0.999 : +0.054% | 
bitops-3bit-bits-in-byte.js | 0.470 -> 0.471 : -0.156% | 
bitops-bits-in-byte.js | 0.645 -> 0.646 : -0.150% | 
bitops-bitwise-and.js | 0.890 -> 0.890 : -0.061% | 
bitops-nsieve-bits.js | 1.128 -> 1.125 : +0.258% | 
controlflow-recursive.js | 0.333 -> 0.332 : +0.549% | 
crypto-aes.js | 0.814 -> 0.809 : +0.644% | 
crypto-md5.js | 0.555 -> 0.556 : -0.098% | 
crypto-sha1.js | 0.542 -> 0.543 : -0.077% | 
date-format-tofte.js | 0.726 -> 0.725 : +0.044% | 
date-format-xparb.js | 0.508 -> 0.513 : -1.076% | 
math-cordic.js | 1.104 -> 1.110 : -0.493% | 
math-partial-sums.js | 0.704 -> 0.701 : +0.372% | 
math-spectral-norm.js | 0.464 -> 0.465 : -0.303% | 
string-base64.js | 1.409 -> 1.368 : **+2.886%** | 
string-fasta.js | 1.307 -> 1.302 : +0.387% | 
Geometric mean: | +0.137% |

Most of it is probably noise due to the new code layout.

Binary (bytes) | 924f4bbd74 | c1ab18398d | Gain |
----: | ----: | ----: | ----: |
size | 132664 | 132664 | 0.000% |
.text | 115596 | 114404 | +1.031% |
.rodata | 11256 | 11262 | -0.053% |
.bss | 1575168 | 1575168 | 0.000% |

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 1]
```diff
@@ -34,6 +34,45 @@
...
+  ECMA_ARRAY_PROTOTYPE_FOR_EACH,
+  ECMA_ARRAY_PROTOTYPE_MAP,
+  ECMA_ARRAY_PROTOTYPE_FILTER,
+  /* Note these 2 routines must be in this order */
+  ECMA_ARRAY_PROTOTYPE_REDUCE,
+  ECMA_ARRAY_PROTOTYPE_REDUCE_RIGHT,
+  ECMA_ARRAY_PROTOTYPE_FIND,
+};
+
+
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: Only one newline.

**Comment by LaszloLango** [comment]
@rerobika I don't see how is this PR related to #2640. Could you explain it to me?

**Comment by rerobika** [comment]
@LaszloLango #2640 Adds 4 new Array.prototype routines [here](https://github.com/pando-project/jerryscript/pull/2640/files#diff-51488696255835eac56729bc7cab1251R70), but currently it uses the general dispatcher. If it lands, this patch must be updated with these 4 functions custom dispatcher related bindings.

**Comment by LaszloLango** [comment]
@rerobika as I see it can be done the other way as well.

**Comment by rerobika** [comment]
Indeed, no matter what PR lands first, the other must be updated as well.

**Comment by rerobika** [comment]
@LaszloLango I've updated the PR after #2640 has landed. Could you take an other look at it?

**Comment by LaszloLango** [comment]
@rerobika, are measurement results the same after the rebase?

**Comment by rerobika** [comment]
@LaszloLango Yes they are, these are the new results:

Binary (bytes) | master(f22eea5c6e) | patch(cec478847b) | Gain |
----: | ----: | ----: | ----: |
size | 132664 | 132664 | 0 bytes |
.text | 115636 | 114444 | -1192 bytes |
.rodata | 11256 | 11255 | -1 bytes |
.bss | 1575168 | 1575168 | 0 bytes |

Performance is still unchanged, mostly noise due to the new code layout.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 2]
```diff
@@ -2497,124 +2059,277 @@ ecma_builtin_array_prototype_object_find (ecma_value_t this_arg, /**< this argum
...
+      ret_value = ecma_builtin_array_prototype_object_find (routine_arg_1,
+                                                            routine_arg_2,
+                                                            obj_p,
+                                                            length);
+      break;
+    }
+#endif /* !CONFIG_DISABLE_ES2015_BUILTIN */
+    default:
+    {
+      JERRY_UNREACHABLE ();
```
Comment (LaszloLango) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: AFAIK this is a bad practice and we started to refactor these unnecessary `JERRY_UNREACHABLE` calls. Please check #2342 for details. CC: @akosthekiss

Comment (rerobika) [reply jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: Sure, it can be simply eliminated.

**Pull request merged**